### PR TITLE
add loadbalancer service in ovn-default vpc

### LINKED ISSUE #1470: Loadbalancer type Service
### Feature request

We already have underlay and overlay mixed cluster, if one pod has both overlay and underlay network it can act as the gateway between two types of network which can be used to implement LoadBalancer type Service.

### PR DESCRIPTION
**Pull request by hongzhen-ma**
#### What type of this PR
Add support for loadbalancer service 

/kind feature

#### Which issue(s) this PR fixes:
Fixes #1470

### CODE REVIEW

**Comment by chestack** [comment]
supporting loadbalancer type service is definitely a helpful feature. It will be great to add some design doc, maybe chinese doc could go first. Thanks.

File: pkg/controller/service_lb.go [hunk 1]
```diff
@@ -0,0 +1,396 @@
...
+	k8serrors "k8s.io/apimachinery/pkg/api/errors"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	utilruntime "k8s.io/apimachinery/pkg/util/runtime"
+	"k8s.io/klog/v2"
+
+	"github.com/kubeovn/kube-ovn/pkg/util"
+)
+
+const (
+	LbPodImage       = "kubeovn/vpc-nat-gateway:v1.11.0"
```
Comment (oilbeater) [inline pkg/controller/service_lb.go:23]: Image name should be configurable, not all users can download image from public

File: pkg/controller/service_lb.go [hunk 2]
```diff
@@ -0,0 +1,396 @@
...
+	POD_EIP_ADD      = "eip-add"
+	POD_DNAT_ADD     = "dnat-add"
+)
+
+func genLbSvcDpName(name string) string {
+	return fmt.Sprintf("lb-svc-%s", name)
+}
+
+func (c *Controller) applyExternalNetwork(svc *corev1.Service) error {
+	nic := "eth0"
```
Comment (oilbeater) [inline pkg/controller/service_lb.go:34]: nic should be configurable

**Review by oilbeater — CHANGES_REQUESTED** [review]

File: Makefile [hunk 3]
```diff
@@ -298,11 +298,14 @@ kind-install-multus:
 	if ! docker images --format "{{.Repository}}:{{.Tag}}" | grep -qw "^$(MULTUS_IMAGE)$$"; then \
 		docker pull "$(MULTUS_IMAGE)"; \
 	fi
+	docker pull kubeovn/vpc-nat-gateway:v1.11.0
```
[the remark this answers is not in the document]
Comment (hongzhen-ma) [reply Makefile]: modified

File: pkg/controller/service_lb.go (hunk 1 above)
Comment (hongzhen-ma) [reply pkg/controller/service_lb.go]: add new annotation "ovn.kubernetes.io/lb_svc_img" to configure image

File: pkg/controller/service_lb.go (hunk 2 above)
Comment (hongzhen-ma) [reply pkg/controller/service_lb.go:36]: The annotation "ovn.kubernetes.io/attchmentnic" is used to config nic, the eth0 is used as default value

File: Makefile [hunk 4]
```diff
@@ -298,11 +300,14 @@ kind-install-multus:
 	if ! docker images --format "{{.Repository}}:{{.Tag}}" | grep -qw "^$(MULTUS_IMAGE)$$"; then \
 		docker pull "$(MULTUS_IMAGE)"; \
 	fi
+	docker pull $(VPC_NAT_GW_IMG)
```
Comment (oilbeater) [inline Makefile:303]: should use the image build locally

**Review by oilbeater — CHANGES_REQUESTED** [review]

File: pkg/controller/service_lb.go [hunk 5]
```diff
@@ -0,0 +1,392 @@
...
+			Create(context.Background(), newDp, metav1.CreateOptions{})
+
+		if err != nil {
+			klog.Errorf("failed to create deployment %s, err: %v", newDp.Name, err)
+			return err
+		}
+		return nil
+	} else {
+		deploy = c.updateLbSvcDeployment(svc, deploy)
+		_, err := c.config.KubeClient.AppsV1().Deployments(svc.Namespace).Update(context.Background(), deploy, metav1.UpdateOptions{})
```
Comment (oilbeater) [inline pkg/controller/service_lb.go:170]: if _, err := ...; err != nil {}

File: pkg/controller/service_lb.go [hunk 6]
```diff
@@ -0,0 +1,392 @@
...
+	return nil
+}
+
+func (c *Controller) getLbSvcPod(svcName, svcNamespace string) (*corev1.Pod, error) {
+	sel, _ := metav1.LabelSelectorAsSelector(&metav1.LabelSelector{
+		MatchLabels: map[string]string{"app": genLbSvcDpName(svcName), "namespace": svcNamespace},
+	})
+
+	pods, err := c.podsLister.Pods(svcNamespace).List(sel)
+	if err != nil {
```
Comment (oilbeater) [inline pkg/controller/service_lb.go:186]: should loop here to wait ready

File: pkg/controller/service_lb.go [hunk 7]
```diff
@@ -0,0 +1,392 @@
...
+		loadBalancerIP = pod.Annotations[attchIpAnnotation]
+	} else {
+		err = fmt.Errorf("failed to get attachment ip from pod's annotation")
+	}
+
+	return loadBalancerIP, err
+}
+
+func (c *Controller) deleteLbSvc(svc *corev1.Service) error {
+	if _, err := c.config.KubeClient.AppsV1().Deployments(svc.Namespace).Get(context.Background(), genLbSvcDpName(svc.Name), metav1.GetOptions{}); err != nil {
```
Comment (oilbeater) [inline pkg/controller/service_lb.go:243]: We can just delete here and check the error

File: pkg/controller/service_lb.go [hunk 8]
```diff
@@ -0,0 +1,392 @@
...
+			return err
+		}
+	}
+
+	if err := c.config.KubeClient.AppsV1().Deployments(svc.Namespace).Delete(context.Background(), genLbSvcDpName(svc.Name), metav1.DeleteOptions{}); err != nil {
+		klog.Errorf("failed to delete deployment %s, err: %v", genLbSvcDpName(svc.Name), err)
+		return err
+	}
+
+	for {
```
Comment (oilbeater) [inline pkg/controller/service_lb.go:256]: no need to wait pod deletion

File: pkg/controller/service_lb.go [hunk 9]
```diff
@@ -0,0 +1,392 @@
...
+	}
+	return true
+}
+
+func (c *Controller) handleUpdateLbSvcPod(svc *corev1.Service) error {
+	pod, err := c.getLbSvcPod(svc.Name, svc.Namespace)
+	if err != nil {
+		// There's no other way to delete update pod process in queue,so record err and return if pod does not exist
+		klog.Errorf("failed to get lb svc pod, %v", err)
+		return nil
```
Comment (oilbeater) [inline pkg/controller/service_lb.go:384]: should requeue here

**Review by oilbeater — CHANGES_REQUESTED** [review]

File: pkg/controller/service_lb.go [hunk 10]
```diff
@@ -0,0 +1,392 @@
...
+
+	providerName := getAttachNetworkProvider(svc)
+	values := strings.Split(providerName, ".")
+	if len(values) <= 1 {
+		return attachmentName, attachmentNs, fmt.Errorf("The provider name should be consisted of name and namesapce")
+	}
+	attachmentName = values[0]
+	attachmentNs = values[1]
+
+	networkClient := c.config.AttachNetClient.K8sCniCncfIoV1().NetworkAttachmentDefinitions(attachmentNs)
```
[the remark this answers is not in the document]
Comment (hongzhen-ma) [reply pkg/controller/service_lb.go]: modified

File: pkg/controller/service_lb.go [hunk 11]
```diff
@@ -0,0 +1,392 @@
...
+		"namespace": svc.Namespace,
+		"service":   svc.Name,
+	}
+
+	image := "kubeovn/vpc-nat-gateway:v1.11.0"
+	if svc.Annotations[util.LbSvcPodImg] != "" {
+		image = svc.Annotations[util.LbSvcPodImg]
+	}
+
+	attachmentName, attachmentNs, _ := c.checkAttachNetwork(svc)
```
[the remark this answers is not in the document]
Comment (hongzhen-ma) [reply pkg/controller/service_lb.go]: modified

File: pkg/controller/service_lb.go [hunk 12]
```diff
@@ -0,0 +1,392 @@
...
+	if needToCreate {
+		newDp := c.genLbSvcDeployment(svc)
+		_, err := c.config.KubeClient.AppsV1().Deployments(svc.Namespace).
+			Create(context.Background(), newDp, metav1.CreateOptions{})
+
+		if err != nil {
+			klog.Errorf("failed to create deployment %s, err: %v", newDp.Name, err)
+			return err
+		}
+		return nil
```
[the remark this answers is not in the document]
Comment (hongzhen-ma) [reply pkg/controller/service_lb.go]: modified

File: pkg/controller/service_lb.go (hunk 6 above)
Comment (hongzhen-ma) [reply pkg/controller/service_lb.go:187]: the loop exists where the function is called

File: pkg/controller/service_lb.go (hunk 7 above)
Comment (hongzhen-ma) [reply pkg/controller/service_lb.go]: modified

File: pkg/controller/service_lb.go (hunk 8 above)
Comment (hongzhen-ma) [reply pkg/controller/service_lb.go]: Modified.
This is used when the net-attach-def is created for each pod, and delete net-attach-def after pod is deleted.

File: pkg/controller/service_lb.go (hunk 9 above)
Comment (hongzhen-ma) [reply pkg/controller/service_lb.go]: modified. 
The new queue has been deleted and processed in create svc function.

**Review by oilbeater — APPROVED** [review]

**Pull request merged**